### PR TITLE
[Attention block] relative positional embedding

### LINKED ISSUE #7356: Self attention with relative positional embedding
**Is your feature request related to a problem? Please describe.**
Enable relative positional embedding in attention block

**Describe the solution you'd like**
A constructor argument specifying which type (or None) of relative embedding to use

**Describe alternatives you've considered**
NA

**Additional context**
Should be useful for some network implementation like SAM (https://arxiv.org/abs/2304.02643, #6357 )
PR opened #7346 with suggestion for decomposed relative positional embedding (https://arxiv.org/abs/2112.01526)

### PR DESCRIPTION
**Pull request by vgrau98**
Fixes #7356

### Description

Add relative positinoal embedding in attention block as described in https://arxiv.org/pdf/2112.01526.pdf
Largely inspired by https://github.com/facebookresearch/segment-anything/blob/main/segment_anything/modeling/image_encoder.py

Can be useful for #6357

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [ ] Quick tests passed locally by running `./runtests.sh --quick --unittests  --disttests`.
- [x] In-line docstrings updated.
- [x] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

**Comment by marksgraham** [comment]
I wonder if this is an example of the point we should start splitting out different types of attention block into their own blocks? Thoughts @KumoLiu  @ericspod ?

Note this also will have some merge conflicts as it stands with the [local window pr](https://github.com/Project-MONAI/MONAI/pull/7348)

**Comment by ericspod** [comment]
@marksgraham @KumoLiu I feel with the generative merging in progress we should think about refactoring how these blocks work in general to try and separate these differing features into separate classes. We can perhaps have a single attention block class that is customisable with features, but if not we can do multiple classes but doing this cleanly will probably involve refactoring the existing things that have been merged from GenerativeModels. This PR adds something good however so we should keep this open but postpone until we have the other things merged perhaps.

**Comment by vgrau98** [comment]
Following @ericspod  suggestions:
Use factory pattern to implement relative positional embedding in self attention block (use `LayerFactory` for now but open to any suggestion to better organise code). Anyone can extend the  `RelPosEmbedding` feature, or add new transformer/attention features

File: monai/networks/blocks/rel_pos_embedding.py
Comment (KumoLiu) [inline monai/networks/blocks/rel_pos_embedding.py:20]: Decoupled the posembedding outside looks cleaner to me. We could also consider moving other attention components as a block such as 
https://github.com/Project-MONAI/MONAI/blob/c61c6ac2d56af08fbbfb955324b8639e266a25db/monai/networks/nets/transformer.py#L116
https://github.com/Project-MONAI/MONAI/blob/c61c6ac2d56af08fbbfb955324b8639e266a25db/monai/networks/nets/transformer.py#L138

File: monai/networks/blocks/selfattention.py
Comment (KumoLiu) [inline monai/networks/blocks/selfattention.py:95]: Agreed with Eric, we should consider how to refactor these blocks. Not sure whether these parts can be used together, if yes, perhaps we should add them in one block otherwise we may add a base class to always avoid large classes.
https://github.com/Project-MONAI/MONAI/blob/c61c6ac2d56af08fbbfb955324b8639e266a25db/monai/networks/nets/transformer.py#L116

Do you have some suggestions here?
cc @ericspod @marksgraham @vgrau98

**Review by KumoLiu — COMMENTED** [review]
Thanks for the quick update, looks cleaner now. Only have a few concern inline.

File: monai/networks/blocks/selfattention.py
Comment (marksgraham) [reply monai/networks/blocks/selfattention.py:95]: the use of `flash attention` does not look compatible with relative positional embeddings to me, as RPE requires access to the attention matrix, whilst flash attention calculates this behind the scenes. but I think we could just check if the user has selected both and throw a warning and/or deactive flash attention if that is the case. torch 2.0 now has a native implementation of flash attention [here](https://pytorch.org/docs/stable/generated/torch.nn.functional.scaled_dot_product_attention.html) so we should move to that when we can, too, rather than rely on an external library.

i believe the `causal` option should be compatible with RPE - we just set some of the attention matrix elements to `-inf` after adding the RPE to it. do you agree @vgrau98?

File: monai/networks/layers/utils.py
Comment (ericspod) [inline monai/networks/layers/utils.py:143]:
```suggestion
def get_rel_pos_embedding_layer(name: tuple | str, s_input_dims: Optional[tuple], c_dim: int, num_heads: int):
    embedding_name, embedding_args = split_args(name)
    embedding_type = RelPosEmbedding[embedding_name]
    # create a dictionary with the default values which can be overridden by embedding_args
    kw_args = {"s_input_dims": s_input_dims, "c_dim": c_dim, "num_heads": num_heads, **embedding_args}
    # filter out unused argument names
    kw_args = {k: v for k, v in kw_args.items() if has_option(embedding_type, k)}
            
    return embedding_type(**kw_args)
```
I think this does what you wanted your implementation to do, unless I'm mistaken there was some issue with your logic when you're checking if a value is *not* in `kw_args` and adding it to `embedding_args` if that was true. In this case you wouldn't use the value `kw_args` if it was present.

File: monai/networks/blocks/attention_utils.py
Comment (ericspod) [inline monai/networks/blocks/attention_utils.py:55]: It's explained in the paper but a short explanation here of what this is would help.

File: monai/networks/blocks/rel_pos_embedding.py
Comment (ericspod) [inline monai/networks/blocks/rel_pos_embedding.py:39]: When would `self.s_input_dims` ever be `None`? If never then this should be enforced in the constructor with a check that raises `ValueError`.

File: monai/networks/blocks/rel_pos_embedding.py
Comment (ericspod) [inline monai/networks/blocks/rel_pos_embedding.py:41]: You could instead do something like this which will set `d` to 1 if `self.input_dims` has 2 values :
```
h, w, d = ensure_tuple_size(self.s_input_dims, 3, 1)
```

File: monai/networks/layers/factories.py
Comment (ericspod) [inline monai/networks/layers/factories.py:478]:
```suggestion
def decomposed_rel_pos_embedding() -> type[nn.Module}:
```

File: monai/networks/blocks/selfattention.py
Comment (ericspod) [reply monai/networks/blocks/selfattention.py:95]: I agree, we can have a check for Pytorch version and use the included function if version 2.0 or higher.

**Comment by ericspod** [comment]
I think this is looking much better. Would it be possible to merge into `gen-ai-dev` instead of `dev` so we can continue the generative refactoring? @KumoLiu @marksgraham any other feedback?

File: monai/networks/blocks/selfattention.py
Comment (vgrau98) [reply monai/networks/blocks/selfattention.py:95]: i agree @marksgraham

File: monai/networks/layers/utils.py
Comment (vgrau98) [reply monai/networks/layers/utils.py]: Yes, you're absolutely right. I commit your suggestion

File: monai/networks/blocks/rel_pos_embedding.py
Comment (vgrau98) [reply monai/networks/blocks/rel_pos_embedding.py]: Never, check added in https://github.com/Project-MONAI/MONAI/pull/7346/commits/2d76c8c0926272a0711e62054bd781235ff5bf4a

File: monai/networks/blocks/rel_pos_embedding.py
Comment (vgrau98) [reply monai/networks/blocks/rel_pos_embedding.py:46]: https://github.com/Project-MONAI/MONAI/pull/7346/commits/ff18e81c5ff610c139383fce3cd415ca39b314b0

File: monai/networks/blocks/attention_utils.py
Comment (vgrau98) [reply monai/networks/blocks/attention_utils.py]: https://github.com/Project-MONAI/MONAI/pull/7346/commits/033ebf82873c5e67dd2798e74468272665725967

Comment (vgrau98) [reply monai/networks/blocks/attention_utils.py]: A simple comment, looking in more detail, if I'm not mistaken the implementation is a little different from the paper definition. 
From https://github.com/facebookresearch/segment-anything/blob/6fdee8f2727f4506cfbbe553e23b895e27956588/segment_anything/modeling/image_encoder.py#L347-L361
It seems we get 
$Attn_{logits} = QK^{T}*scale + E^{(rel)}$
instead of (from paper https://arxiv.org/abs/2112.01526):
$Attn_{logits} = (QK^{T} + E^{(rel)})*scale$

In the implementation, RPE does not seem to be scaled

**Comment by vgrau98** [comment]
> ther feedback

Waiting for @marksgraham confirmation

**Comment by vgrau98** [comment]
> I think this is looking much better. Would it be possible to merge into `gen-ai-dev` instead of `dev` so we can continue the generative refactoring? @KumoLiu @marksgraham any other feedback?

Waiting for @marksgraham confirmation

**Comment by vgrau98** [comment]
> I think this is looking much better. Would it be possible to merge into `gen-ai-dev` instead of `dev` so we can continue the generative refactoring? @KumoLiu @marksgraham any other feedback?

target branch updated to `gen-ai-dev`

**Comment by KumoLiu** [comment]
Hi @vgrau98, could you please try `git rebase --onto  <newparent> <oldparent> <until>` to rebase your commit onto the gen-ai-dev, which may avoid merging the commit added from the dev branch (causing DCO error)?
https://womanonrails.com/git-rebase-onto

**Comment by vgrau98** [comment]
> Hi @vgrau98, could you please try `git rebase --onto <newparent> <oldparent> <until>` to rebase your commit onto the gen-ai-dev, which may avoid merging the commit added from the dev branch (causing DCO error)? https://womanonrails.com/git-rebase-onto

should be ok now

**Review by KumoLiu — APPROVED** [review]

**Comment by vgrau98** [comment]
> Hi @vgrau98, could you please check-pick the commit [here](https://github.com/Project-MONAI/MONAI/pull/7391) after it merged which caused a CI error and could not merge? Thanks.

Is it still relevant? The bug seems to have been resolved according to your comments on PR #7391

**Comment by KumoLiu** [comment]
Yes, I'm creating another PR to fix the error caused by the upgrade of flake8-bugbear
https://github.com/Project-MONAI/MONAI/actions/runs/7544331654/job/20537411217#step:2:2601

You can either cherry-pick that one or make the same change in this PR.

**Pull request merged**
